File: dart_sample/lib/demos/general/state_storage_demo_page.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

class StateStorageDemoPage extends StatefulWidget {
  const StateStorageDemoPage({super.key});

  @override
  State<StateStorageDemoPage> createState() => _StateStorageDemoPageState();
}

class _StateStorageDemoPageState extends State<StateStorageDemoPage> {
  static const String _sharedCounterKey = 'shared-counter';
  final PageStorageBucket _bucket = PageStorageBucket();
  bool _showScrollable = true;

  @override
  Widget build(BuildContext context) {
    return PageStorage(
      bucket: _bucket,
      child: SharedAppData(child: Builder(builder: _buildContent)),
    );
  }

  Widget _buildContent(BuildContext context) {
    final int sharedCounter = SharedAppData.getValue<String, int>(
      context,
      _sharedCounterKey,
      () => 0,
    );
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 12,
      children: <Widget>[
        const Text(
          'PageStorage + SharedAppData',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Jump the list, unmount it, then restore it. The same PageStorageKey '
          'restores the offset; the shared counter rebuilds only its keyed dependent. '
          'The list inherits its controller through PrimaryScrollController and '
          'its desktop chrome through ScrollConfiguration. Drag past an edge to '
          'compare Flutter glow and stretch indicators.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            SizedBox(
              width: 160,
              child: OutlinedButton(
                onPressed: () => SharedAppData.setValue<String, int>(
                  context,
                  _sharedCounterKey,
                  sharedCounter + 1,
                ),
                child: Text('Shared value: $sharedCounter'),
              ),
            ),
            SizedBox(
              width: 160,
              child: OutlinedButton(
                onPressed: () => setState(() {
                  _showScrollable = !_showScrollable;
                }),
                child: Text(_showScrollable ? 'Unmount list' : 'Restore list'),
              ),
            ),
          ],
        ),
        Expanded(
          child: _showScrollable
              ? const _RestorableStorageList()
              : Container(
                  color: const Color(0xFFE8EEF6),
                  alignment: Alignment.center,
                  child: const Text(
                    'List is unmounted. Restore it to verify the saved offset.',
                    style: TextStyle(color: Color(0xFF31506F)),
                  ),
                ),
        ),
      ],
    );
  }
}

class _RestorableStorageList extends StatefulWidget {
  const _RestorableStorageList();

  @override
  State<_RestorableStorageList> createState() => _RestorableStorageListState();
}

class _RestorableStorageListState extends State<_RestorableStorageList> {
  final ScrollController _controller = ScrollController();
  bool _showScrollbar = true;
  bool _useStretch = true;

  @override
  Widget build(BuildContext context) {
    final Widget scrollView = SingleChildScrollView(
      key: const PageStorageKey<String>('state-storage-list'),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        spacing: 6,
        children: List<Widget>.generate(18, _buildRow),
      ),
    );
    final Widget indicatedScrollView = _useStretch
        ? StretchingOverscrollIndicator(
            axisDirection: AxisDirection.down,
            child: scrollView,
          )
        : GlowingOverscrollIndicator(
            axisDirection: AxisDirection.down,
            color: const Color(0xFF625B71),
            child: scrollView,
          );

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 8,
      children: <Widget>[
        Wrap(
          spacing: 8,
          runSpacing: 8,
          children: <Widget>[
            SizedBox(
              width: 180,
              child: FilledButton(
                onPressed: () => _controller.jumpTo(240),
                child: const Text('Jump to offset 240'),
              ),
            ),
            SizedBox(
              width: 180,
              child: FilledButton(
                onPressed: () => setState(() {
                  _showScrollbar = !_showScrollbar;
                }),
                child: Text(
                  _showScrollbar
                      ? 'Hide config scrollbar'
                      : 'Show config scrollbar',
                ),
              ),
            ),
            SizedBox(
              width: 180,
              child: FilledButton(
                onPressed: () => setState(() {
                  _useStretch = !_useStretch;
                }),
                child: Text(_useStretch ? 'Effect: stretch' : 'Effect: glow'),
              ),
            ),
          ],
        ),
        Expanded(
          child: ScrollConfiguration(
            behavior: const _DesktopDemoScrollBehavior().copyWith(
              scrollbars: _showScrollbar,
              dragDevices: <PointerDeviceKind>{
                PointerDeviceKind.touch,
                PointerDeviceKind.mouse,
                PointerDeviceKind.trackpad,
              },
            ),
            child: PrimaryScrollController(
              automaticallyInheritForPlatforms: const <TargetPlatform>{
                TargetPlatform.windows,
              },
              controller: _controller,
              child: indicatedScrollView,
            ),
          ),
        ),
      ],
    );
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  static Widget _buildRow(int index) {
    return Container(
      height: 44,
      color: index.isEven ? const Color(0xFFF4F7FA) : const Color(0xFFE6EDF5),
      padding: const EdgeInsets.symmetric(horizontal: 12),
      alignment: Alignment.centerLeft,
      child: Text('Stored row ${index + 1}'),
    );
  }
}

class _DesktopDemoScrollBehavior extends ScrollBehavior {
  const _DesktopDemoScrollBehavior();

  @override
  TargetPlatform getPlatform(BuildContext context) => TargetPlatform.windows;
}

File: src/Sample/Plumix.Sample/Demos/General/StateStorageDemoPage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Avalonia;
using Avalonia.Media;
using Plumix.Foundation;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source: dart_sample/lib/demos/general/state_storage_demo_page.dart (exact sample parity)

public sealed class StateStorageDemoPage : StatefulWidget
{
    public override State CreateState() => new StateStorageDemoPageState();
}

internal sealed class StateStorageDemoPageState : State
{
    private const string SharedCounterKey = "shared-counter";
    private readonly PageStorageBucket _bucket = new();
    private bool _showScrollable = true;

    public override Widget Build(BuildContext context)
    {
        return new PageStorage(
            _bucket,
            new SharedAppData(
                new Builder(BuildContent)));
    }

    private Widget BuildContent(BuildContext context)
    {
        int sharedCounter = SharedAppData.GetValue(context, SharedCounterKey, () => 0);
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 12,
            children:
            [
                new Text("PageStorage + SharedAppData", fontSize: 20, color: Colors.Black),
                new Text(
                    "Jump the list, unmount it, then restore it. The same PageStorageKey restores the offset; " +
                    "the shared counter rebuilds only its keyed dependent. The list inherits its controller " +
                    "through PrimaryScrollController and its desktop chrome through ScrollConfiguration.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton(
                            $"Shared value: {sharedCounter}",
                            () => SharedAppData.SetValue(context, SharedCounterKey, sharedCounter + 1)),
                        BuildButton(
                            _showScrollable ? "Unmount list" : "Restore list",
                            () => SetState(() => _showScrollable = !_showScrollable)),
                    ]),
                new Expanded(
                    child: _showScrollable
                        ? new RestorableStorageList()
                        : new Container(
                            color: Color.Parse("#FFE8EEF6"),
                            alignment: Alignment.Center,
                            child: new Text(
                                "List is unmounted. Restore it to verify the saved offset.",
                                color: Color.Parse("#FF31506F")))),
            ]);
    }

    private static Widget BuildButton(string label, Action onTap)
    {
        return new SizedBox(
            width: 160,
            child: new CounterTapButton(
                label: label,
                onTap: onTap,
                background: Color.Parse("#FFDCE3ED"),
                foreground: Colors.Black,
                fontSize: 12,
                padding: new Thickness(8, 7)));
    }
}

internal sealed class RestorableStorageList : StatefulWidget
{
    public override State CreateState() => new RestorableStorageListState();
}

internal sealed class RestorableStorageListState : State
{
    private readonly ScrollController _controller = new();
    private bool _showScrollbar = true;

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 8,
            children:
            [
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("Jump to offset 240", () => _controller.JumpTo(240)),
                        BuildButton(
                            _showScrollbar ? "Hide config scrollbar" : "Show config scrollbar",
                            () => SetState(() => _showScrollbar = !_showScrollbar)),
                    ]),
                new Expanded(
                    child: new ScrollConfiguration(
                        behavior: new DesktopDemoScrollBehavior().CopyWith(
                            scrollbars: _showScrollbar,
                            dragDevices: new HashSet<PointerDeviceKind>
                            {
                                PointerDeviceKind.Touch,
                                PointerDeviceKind.Mouse,
                                PointerDeviceKind.Trackpad,
                            }),
                        child: new PrimaryScrollController(
                            controller: _controller,
                            automaticallyInheritForPlatforms:
                                new HashSet<TargetPlatform> { TargetPlatform.Windows },
                            child: new SingleChildScrollView(
                                key: new PageStorageKey<string>("state-storage-list"),
                                child: new Column(
                                    crossAxisAlignment: CrossAxisAlignment.Stretch,
                                    spacing: 6,
                                    children: Enumerable.Range(0, 18)
                                        .Select(BuildRow)
                                        .ToArray()))))),
            ]);
    }

    public override void Dispose()
    {
        _controller.Dispose();
        base.Dispose();
    }

    private static Widget BuildRow(int index)
    {
        return new Container(
            height: 44,
            color: index % 2 == 0 ? Color.Parse("#FFF4F7FA") : Color.Parse("#FFE6EDF5"),
            padding: new Thickness(12, 0),
            alignment: Alignment.CenterLeft,
            child: new Text($"Stored row {index + 1}", color: Colors.Black));
    }

    private static Widget BuildButton(string label, Action onTap)
    {
        return new SizedBox(
            width: 180,
            child: new CounterTapButton(
                label: label,
                onTap: onTap,
                background: Color.Parse("#FF31506F"),
                foreground: Colors.White,
                fontSize: 12,
                padding: new Thickness(8, 7)));
    }
}

internal sealed class DesktopDemoScrollBehavior : ScrollBehavior
{
    public override TargetPlatform GetPlatform(BuildContext context) => TargetPlatform.Windows;
}
